src/contracts/functions/crud.ts

```diff
@@ -0,0 +1,52 @@
+import type {ContractState, GetInput, PutInput, RemoveInput, UpdateInput} from '../types';
+import {KeyExistsError} from '../errors';
+import {assertWhitelist, safeGet, verifyAuthProof} from '../utils';
+
+export async function get<State extends ContractState>(_: State, {key}: GetInput['value']) {
+  return {
+    result: await SmartWeave.kv.get(key),
+  };
+}
+
+export async function put<State extends ContractState<{whitelists: ['put']; circuits: []}>>(
+  state: State,
+  {key, value}: PutInput['value'],
+  caller: string
+) {
+  assertWhitelist(state, caller, 'put');
+
+  if ((await SmartWeave.kv.get(key)) !== null) {
+    throw KeyExistsError;
+  }
+  await SmartWeave.kv.put(key, value);
+
+  return {state};
+}
+
+export async function remove<State extends ContractState<{whitelists: ['update']; circuits: ['auth']}>>(
```

Comment:
by specifying a specific state to extend, we can make functions "safe" to be used in an SDK, for example this function requires an `update` whitelist and an `auth` circuit to exist in the contract state.